### PR TITLE
Distributed groupby sum operation

### PR DESCRIPTION
**Pull request by kayibal**
This was a difficult one...but crucial step to scaling our binning algorithm. Finally it seems to be working also fixed a little bug along the way.

It works roughly like this:
1) apply groupy_sum on each partition
2) hash index and partition into n unique (index) partitions (so each index hashed value is in it"s own partition)
3) concat these mini partitions into number of desired partitions 
4) finally apply groupby_sum on newly formed partitions

This should fix the trouble we had with huge test datasets of a single partitions. So we can predict with dask_xgboost now. it also should allow us to train on much more data. Given we have a cluster.

### CODE REVIEW

File: sparsity/dask/core.py [hunk 1]
```diff
@@ -518,4 +528,179 @@ def apply_and_enforce(func, arg, kwargs, meta):
...
+    else:
+        dsk = {(a, 0, i, 0): (apply, chunk,
+                              [(x._name, i) if isinstance(x, SparseFrame)
+                               else x for x in args], chunk_kwargs)
+               for i in range(args[0].npartitions)}
+
+    # Split
+    # this splits the blocks (usually) by their index and
+    # basically performs a task sort such that the next tree
+    # aggregation will result in the disered number of partitions
```
Comment (vitords) [inline sparsity/dask/core.py]: (...) desired (...)

File: sparsity/dask/core.py [hunk 2]
```diff
@@ -518,4 +528,179 @@ def apply_and_enforce(func, arg, kwargs, meta):
...
+        split_prefix = 'split-%s' % token_key
+        shard_prefix = 'shard-%s' % token_key
+        for i in range(args[0].npartitions):
+            # For now we assime that split_out_setup selects the index
+            # as we will only support index groupbys for now. So we can
+            # use the function provided by dask.
+            dsk[(split_prefix, i)] = (hash_shard, (a, 0, i, 0), split_out,
+                                      split_out_setup, split_out_setup_kwargs)
+            # at this point we have dictionaries of dataframes the key
+            # is the to which output partition this dataframe belongs to
```
Comment (vitords) [inline sparsity/dask/core.py]: Confusing comment, the key is the what?

File: sparsity/dask/core.py [hunk 3]
```diff
@@ -518,4 +528,179 @@ def apply_and_enforce(func, arg, kwargs, meta):
...
+    # Split
+    # this splits the blocks (usually) by their index and
+    # basically performs a task sort such that the next tree
+    # aggregation will result in the disered number of partitions
+    # given by the split_out parameter
+    if split_out and split_out > 1:
+        split_prefix = 'split-%s' % token_key
+        shard_prefix = 'shard-%s' % token_key
+        for i in range(args[0].npartitions):
+            # For now we assime that split_out_setup selects the index
```
Comment (vitords) [inline sparsity/dask/core.py]: (...) assume (...)

File: sparsity/dask/core.py (hunk 1 above)
Comment (vitords) [inline sparsity/dask/core.py]: (...) desired (...)

File: sparsity/dask/core.py [hunk 4]
```diff
@@ -518,4 +528,179 @@ def apply_and_enforce(func, arg, kwargs, meta):
...
+    b = '{0}-combine-{1}'.format(token or funcname(combine), token_key)
+    k = npartitions
+    depth = 0
+    while k > split_every:
+        for part_i, inds in enumerate(partition_all(split_every, range(k))):
+            for j in range(split_out):
+                # concat basically just concatenates partitions along
+                # the first dimension. Will replace this with the vstack
+                # method from the SparseFrame class. Dask does some
+                # extra checks which we can hopefully neglect.
```
Comment (vitords) [inline sparsity/dask/core.py]: I think this comment is not necessary anymore as it's already using `vstack`.
Dask checks can be in fact neglected? (I have no idea which checks are these)

Comment (kayibal) [reply sparsity/dask/core.py]: Ok removed the comment. Dask does some extra checks at this point e.g. if the columns are the same but vstack does that too.

File: sparsity/dask/core.py (hunk 2 above)
Comment (kayibal) [reply sparsity/dask/core.py]: Tried to explain it a bit better....

**Comment by kayibal** [comment]
Too magical oO

**Pull request merged**